packages/block-editor/src/components/block-actions/index.js

```diff
@@ -6,129 +6,106 @@ import { castArray, first, last, every } from 'lodash';
 /**
  * WordPress dependencies
  */
-import { compose } from '@wordpress/compose';
-import { withSelect, withDispatch } from '@wordpress/data';
+import { useDispatch, useSelect } from '@wordpress/data';
 import { hasBlockSupport, switchToBlockType } from '@wordpress/blocks';
 
-function BlockActions( {
-	canDuplicate,
-	canInsertDefaultBlock,
-	children,
-	isLocked,
-	onDuplicate,
-	onGroup,
-	onInsertAfter,
-	onInsertBefore,
-	onRemove,
-	onUngroup,
-	blocks,
-} ) {
+/**
+ * Internal dependencies
+ */
+import { useNotifyCopy } from '../copy-handler';
+
+export default function BlockActions( { clientIds, children } ) {
+	const {
+		canInsertBlockType,
+		getBlockRootClientId,
+		getBlocksByClientId,
+		getTemplateLock,
+	} = useSelect( ( select ) => select( 'core/block-editor' ), [] );
+	const {
+		getDefaultBlockName,
+		getGroupingBlockName,
+	} = useSelect( ( select ) => select( 'core/blocks' ) );
+
+	const blocks = getBlocksByClientId( clientIds );
+	const rootClientId = getBlockRootClientId( clientIds[ 0 ] );
+	const canDuplicate = every( blocks, ( block ) => {
+		return (
+			!! block &&
+			hasBlockSupport( block.name, 'multiple', true ) &&
+			canInsertBlockType( block.name, rootClientId )
+		);
+	} );
+
+	const canInsertDefaultBlock = canInsertBlockType(
+		getDefaultBlockName(),
+		rootClientId
+	);
+
+	const {
+		removeBlocks,
+		replaceBlocks,
+		duplicateBlocks,
+		insertAfterBlock,
+		insertBeforeBlock,
+		flashBlock,
+	} = useDispatch( 'core/block-editor' );
+
+	const notifyCopy = useNotifyCopy();
+
 	return children( {
 		canDuplicate,
 		canInsertDefaultBlock,
-		isLocked,
-		onDuplicate,
-		onGroup,
-		onInsertAfter,
-		onInsertBefore,
-		onRemove,
-		onUngroup,
+		isLocked: !! getTemplateLock( rootClientId ),
+		rootClientId,
 		blocks,
+		onDuplicate() {
+			return duplicateBlocks( clientIds );
+		},
+		onRemove() {
+			removeBlocks( clientIds );
+		},
+		onInsertBefore() {
+			insertBeforeBlock( first( castArray( clientIds ) ) );
+		},
+		onInsertAfter() {
+			insertAfterBlock( last( castArray( clientIds ) ) );
+		},
+		onGroup() {
+			if ( ! blocks.length ) {
+				return;
+			}
+
+			const groupingBlockName = getGroupingBlockName();
+
+			// Activate the `transform` on `core/group` which does the conversion
+			const newBlocks = switchToBlockType( blocks, groupingBlockName );
+
+			if ( ! newBlocks ) {
+				return;
+			}
+			replaceBlocks( clientIds, newBlocks );
+		},
+		onUngroup() {
+			if ( ! blocks.length ) {
+				return;
+			}
+
+			const innerBlocks = blocks[ 0 ].innerBlocks;
+
+			if ( ! innerBlocks.length ) {
+				return;
+			}
+
+			replaceBlocks( clientIds, innerBlocks );
+		},
+		onCopy() {
+			const selectedBlockClientIds = blocks.map(
+				( { clientId } ) => clientId
+			);
+			if ( blocks.length === 1 ) {
+				flashBlock( selectedBlockClientIds[ 0 ] );
+			}
+			notifyCopy( 'copy', selectedBlockClientIds );
```

Comment:
Have you weighed the benefits of combining _notify_ and _flash_ somewhere — maybe at the action level?

Reply:
Of course I thought about combining them, but since `flashBlock` is used only if `one` block is copied/cut, didn't seem to me so `consistent` with the handling of the same actions with more blocks. So I think that `CopyHandler` might need more thought about this consistency in the future.

With the above said, I believe I need some more experience with the project, to understand it better and feel more comfortable before starting making more design decisions. I'm observing a bit more for now :)

Great thought provoking feedback Miguel! Really appreciate it!